pkg/azuredevops/smoke/smoketests.go

```diff
@@ -0,0 +1,401 @@
+package smoketests
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/microsoft/azure-devops-go-api/azuredevops"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/build"
+	"golang.org/x/exp/slices"
+)
+
+type BuildTest struct {
+	Description  string
+	LogMessage   string
+	ExpectAbsent bool
+}
+
+type TimelineTest struct {
+	Name   string
+	State  string
+	Result string
+}
+
+func CreatePatConnection(organizationURL, personalAccessToken string) *azuredevops.Connection {
+	return azuredevops.NewPatConnection(organizationURL, personalAccessToken)
+}
+
+func getSpecificBuilds(ctx context.Context, connection *azuredevops.Connection, projectName string, pipelineID int) ([]build.Build, error) {
+	buildClient, err := build.NewClient(ctx, connection)
+	if err != nil {
+		return nil, fmt.Errorf("error creating build client: %w", err)
+	}
+
+	// Args to download Build
+	buildArgs := build.GetBuildsArgs{
+		Project:     &projectName,
+		Definitions: &[]int{pipelineID},
+	}
+
+	// Download Builds
+	buildsResponse, err := buildClient.GetBuilds(ctx, buildArgs)
+	if err != nil {
+		return nil, fmt.Errorf("error getting builds: %w", err)
+	}
+
+	return buildsResponse.Value, nil
+}
+
+func getBuildStageStatus(ctx context.Context, connection *azuredevops.Connection, projectName string, buildID int, test TimelineTest) (bool, error) {
+	buildClient, err := build.NewClient(ctx, connection)
+	if err != nil {
+		return false, fmt.Errorf("error creating build client: %w", err)
+	}
+
+	// Args to download Build
+	buildArgs := build.GetBuildTimelineArgs{
+		Project: &projectName,
+		BuildId: &buildID,
+	}
+	// Download Builds
+	buildTimeline, err := buildClient.GetBuildTimeline(ctx, buildArgs)
+	if err != nil {
+		return false, fmt.Errorf("error getting builds: %w", err)
+	}
+
+	return checkBuildRecords(buildTimeline, test.Name, test.Result, test.State)
+}
+
+func checkBuildRecords(timeline *build.Timeline, testName, testResult, testState string) (bool, error) {
+	for _, record := range *timeline.Records {
+		if record.Name != nil && *record.Name == testName {
+			if record.Result != nil && string(*record.Result) == testResult && record.State != nil && string(*record.State) == testState {
+				return true, nil // Found a record matching all criteria
+			}
+		}
+	}
+
+	return false, fmt.Errorf("no record found matching the criteria")
+}
+
+func checkSpecificBuildForCommand(ctx context.Context, connection *azuredevops.Connection, projectName, pipelineName, logFinding string, pipelineID int) (bool, error) {
+	builds, err := getSpecificBuilds(ctx, connection, projectName, pipelineID)
```

Comment:
With calling getSpecificBuilds we have no guarantee that we will be testing the run with changes. It might happen that another pipeline run will finish and will be the last one before we will get the list from API. We should pass a pipeline runID instead getting last run of a pipeline.

Reply:
Reworked. This variable is available as environmental variable, so it was possible to simply pass it to this function.